cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -91,51 +193,601 @@ public void schedule(Id.Program program, SchedulableProgramType programType, Ite
+    }
+
+    @Override
+    public Type getNotificationFeedType() {
+      return StreamSizeNotification.class;
+    }
+
+    @Override
+    public void received(final StreamSizeNotification notification, NotificationContext notificationContext) {
+      // We only use the stream size notification to the schedule tasks if it came after the last seen notification
+      boolean send = false;
+      if (lastNotification == null || notification.getTimestamp() > lastNotification.getTimestamp()) {
+        send = true;
+        lastNotification = notification;
+      }
+      if (send) {
+        if (activeTasks.get() > 0) {
+
+          boolean poll = false;
+          if (delta == null) {
+            poll = true;
+          } else {
+            for (StreamSizeScheduleTask streamSizeScheduleTask : scheduleTasks.values()) {
+              if (streamSizeScheduleTask.shouldTriggerProgram(notification.getSize() - delta)) {
+                poll = true;
+                break;
+              }
+            }
+          }
+
+          if (poll) {
+            pollBookingExecutor.schedule(new Runnable() {
+              @Override
+              public void run() {
+                try {
+                  StreamSize streamSize = pollOnce();
+                  delta = notification.getSize() - streamSize.getSize();
+                  sendPollingInfoToActiveTasks(new StreamSizeNotification(streamSize.getTimestamp(),
+                                                                          streamSize.getSize()));
+                } catch (Throwable t) {
+                  LOG.error("Could not poll stream {}", streamId.getName(), t);
+                }
+              }
+            }, Constants.MetricsCollector.DEFAULT_FREQUENCY_SECONDS, TimeUnit.SECONDS);
+          }
+        }
+      }
+    }
+
+    /**
+     * Add a new scheduling task based on the data received by the stream referenced by {@code this} object.
+     */
+    public void createScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                   StreamSizeSchedule streamSizeSchedule, boolean active,
+                                   long baseRunSize, long baseRunTs, boolean persist) throws SchedulerException {
+      final StreamSizeScheduleTask newTask = new StreamSizeScheduleTask(programId, programType, streamSizeSchedule);
+      synchronized (this) {
+        StreamSizeScheduleTask previous =
+          scheduleTasks.putIfAbsent(AbstractSchedulerService.scheduleIdFor(programId, programType,
+                                                                           streamSizeSchedule.getName()),
+                                    newTask);
+        if (previous != null) {
+          // We cannot replace an existing schedule - that functionality is not wanted - yet
+          throw new SchedulerException("Tried to overwrite schedule " + streamSizeSchedule.getName());
+        }
+
+        if (active) {
+          activeTasks.incrementAndGet();
```

Comment:
Yes you're right. It means that I have to include the task.startSchedule() inside the synchronized block too. I will do that.